File: Engine/graphicsystem/base/RenderToTexture.cc
/****************************************************************************
Copyright (c) 2011-2013,WebJet Business Division,CYOU

http://www.genesis-3d.com.cn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
#include "stdneb.h"
#include "RenderToTexture.h"
#include "GraphicSystem.h"
#if __WIN32__
#include "rendersystem/d3d9/RenderDeviceD3D9.h"
#include "d3d9/TextureD3D9.h"
#include "d3d9/D3D9Types.h"
#endif

#include "addons/resource/imageoperations.h"
#include "Renderable/QuadRenderable.h"

namespace Graphic
{
	__ImplementClass(RenderToTexture,'RTGS',Core::RefCounted)
		__ImplementClass(RenderToTextureManager,'RTMG',Core::RefCounted);
	__ImplementImageSingleton(RenderToTextureManager);

	RenderToTexture::RenderToTexture()
		: m_width(1),
		m_height(1)
	{

	}

	RenderToTexture::~RenderToTexture()
	{
		Discard();
	}

	void RenderToTexture::Setup(const int& width, const int& height, const RenderBase::PixelFormat::Code& colorFormat, const uint& flag, const Math::float4& color, 
		const bool& useDepth, const float& screenRatio,
		RenderBase::AntiAliasQuality::Code quality,
		const Math::rectangle<int>& resolveRect, const bool& isBackBuffer)
	{

		GPtr<RenderBase::RenderTarget> shareDepthStencilRT = m_shareDepthStencilRT;

		// 为了修复共享深度buffer的bug，创建一个m_shareDepthStencilRT的备份
		Discard();
		m_shareDepthStencilRT = shareDepthStencilRT;

		m_RT = RenderBase::RenderTarget::Create();
		m_RT->SetDefaultRenderTarget(isBackBuffer);
		m_RT->SetWidth(width);
		m_RT->SetHeight(height);

		if (resolveRect.right == 0 && resolveRect.left == 0 && resolveRect.top == 0 && resolveRect.bottom == 0)
		{
		}
		else
		{
			m_RT->SetResolveTextureWidth(resolveRect.width());
			m_RT->SetResolveTextureHeight(resolveRect.height());
			m_RT->SetResolveRect(resolveRect);
		}

		m_RT->SetScreenSizeRatio(screenRatio);
		m_RT->SetClearColor(color);
		m_RT->SetClearFlags(flag);
		m_RT->SetColorBufferFormat(colorFormat);
		m_RT->SetAntiAliasQuality(quality);
		if (useDepth)
		{
			if (m_shareDepthStencilRT.isvalid())
			{
				m_RT->AddSharedDepthStencilBuffer(m_shareDepthStencilRT);
			}
			else
			{
				m_RT->AddDepthStencilBuffer();
			}
		}
		m_rtHandle = GraphicSystem::Instance()->CreateRenderTarget(m_RT,m_texHandle);

		if (!m_quadRenderable.isvalid())
		{
			m_quadRenderable = QuadRenderable::Create();
		}
		else
		{
			m_quadRenderable->Discard();
		}

		m_quadRenderable->Setup(width, height);

		m_width  = width;
		m_height = height;
	}

	void RenderToTexture::ChangeSize(int width,int height)
	{
		m_RT->SetWidth(width);
		m_RT->SetHeight(height);
		GraphicSystem::Instance()->ReSizeRenderTarget(m_rtHandle,width,height);

		if (m_quadRenderable.isvalid())
		{
			m_quadRenderable->Discard();
		}
		else
		{
			m_quadRenderable = QuadRenderable::Create();
		}

		m_quadRenderable->Setup(width, height);
	}

	void RenderToTexture::Discard()
	{
		if (m_texHandle.IsValid())
		{
			GraphicSystem::Instance()->RemoveTexture(m_texHandle);
			m_texHandle = dummyhandle;
		}

		if (m_rtHandle.IsValid())
		{
			GraphicSystem::Instance()->RemoveRenderTarget(m_rtHandle);
			m_rtHandle = dummyhandle;
		}
		m_shareDepthStencilRT = 0;
		m_RT = 0;
	}

	bool RenderToTexture::SaveTexture( const Util::String& path )
	{
#if __WIN32__
		Core::RefCounted* Ro = this->m_texHandle.AsObject();
		n_assert(Ro!=NULL);
		D3D9::TextureD3D9 * Instance = reinterpret_cast<D3D9::TextureD3D9*>(Ro);
		n_assert(Instance!=NULL);

		SizeT levelPixelSize =Instance->GetTextureSize();

		ubyte* data = new ubyte[levelPixelSize];
		n_assert(data!=NULL);
		Instance->GetTextureData(data);

		GPtr<Resources::ImageRes> textureImageRes = Resources::ImageRes::Create();
		textureImageRes->SetHeight(Instance->GetHeight());
		textureImageRes->SetWidth(Instance->GetWidth());
		textureImageRes->SetPixelFormat(Instance->GetPixelFormat());
		textureImageRes->SetDepth(1);
		textureImageRes->SetMipMapLevel(1);
		textureImageRes->SetNumFace(1);
		textureImageRes->SetImageData(data, levelPixelSize);
		return Resources::ImageOperations::SaveTGAImage(path, textureImageRes);
#endif

		return false;
	}

#if __WIN32__
	const D3D9::TextureD3D9& RenderToTexture::GetD3DTexture9()
	{
		const RenderBase::TextureHandle handle = this->GetTextureHandle();
		const Core::RefCounted* Ro = handle.AsObject();
		const D3D9::TextureD3D9* front = reinterpret_cast<const D3D9::TextureD3D9*>(Ro);
		return *front;
	}
#endif

	void RenderToTexture::ResetQuad()
	{
		m_quadRenderable->Setup(m_width, m_height);
	}

	RenderToTextureManager::RenderToTextureManager()
	{
		__ConstructImageSingleton
	}

	RenderToTextureManager::~RenderToTextureManager()
	{
		__DestructImageSingleton
			m_RenderToTextures.Clear();
	}


}

File: Engine/script/runtimelibrary/RenderToTexture.cs
/****************************************************************************
Copyright (c) 2011-2013,WebJet Business Division,CYOU
 
http://www.genesis-3d.com.cn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Runtime.CompilerServices;
using ScriptRuntime;
namespace ScriptRuntime
{
    //copy from Engine\rendersystem\base\PixelFormat.h
    internal enum _PixelFormat : int
    {
        // begin common
        X8R8G8B8 = 0,
        R8G8B8,
        B8G8R8,
        B8G8R8A8,
        R8G8B8A8,
        A8B8G8R8,
        A8R8G8B8,
        R5G6B5,
        A1R5G5B5,
        A4R4G4B4,
        DXT1,
        DXT2,
        DXT3,
        DXT4,
        DXT5,
        R16F,                       // 16 bit float, red only
        G16R16F,                    // 32 bit float, 16 bit red, 16 bit green
        A16B16G16R16F,              // 64 bit float, 16 bit rgba each
        R32F,                       // 32 bit float, red only
        G32R32F,                    // 64 bit float, 32 bit red, 32 bit green
        A32B32G32R32F,              // 128 bit float, 32 bit rgba each
        A8,
        L8,
        L8A8,						//[2012/4/12 zhongdaohuan] mygui 用到的格式
        L16,
        G16R16,
        D24X8,
        D24S8,
        // end common   CommonCount

        // begin iPhone /  PowerVR
        PVRTC_RGB2,
        PVRTC_RGBA2,
        PVRTC_RGB4,
        PVRTC_RGBA4,
        // end iphone / PowerVR end   iOSCount

        //ETC
        ETC1_RGB8,

        NumPixelFormats,
        InvalidPixelFormat,
        DescPixelFormatsCount,	//	不调整NumPixelFormats和InvalidPixelFormat的顺序，以免引起BUG。用于像素格式描述的数组

        // 记录分块的标志位。 更改枚举之后，记得修改
        CommonCount = D24S8 + 1,
        iOSCount = PVRTC_RGBA4 + 1,
    };

    public enum PixelFormat : int
    {
        X8R8G8B8 = _PixelFormat.X8R8G8B8,
        R16F = _PixelFormat.R16F,
        G16R16F = _PixelFormat.G16R16F,
        A16B16G16R16F = _PixelFormat.A16B16G16R16F,
        R32F = _PixelFormat.R32F,
        G32R32F = _PixelFormat.G32R32F,
        A32B32G32R32F = _PixelFormat.A32B32G32R32F,

    }

    public enum ClearFlag
    {
        ClearColor = (1 << 0),
        ClearDepth = (1 << 1),
        ClearStencil = (1 << 2),
        ClearAll = ClearColor | ClearDepth | ClearStencil,
    };

    /// <summary>
    /// 渲染到纹理
    /// </summary>
    public class RenderToTexture : Texture
    {
        private RenderToTexture(DummyClass__ dummyObj)
        {

        }

        ~RenderToTexture()
        {
            ICall_RenderToTexture_Release(this);
        }

        /// <summary>
        /// 默认构造函数
        /// </summary>
        public RenderToTexture()
        {
            ICall_RenderToTexture_Bind(this);
        }

        /// <summary>
        /// 创建一个用户自定义的渲染到纹理
        /// </summary>
        /// <param name="width">渲染到纹理的宽度</param>
        /// <param name="height">渲染到纹理的高度</param>
        /// <param name="format">渲染到纹理的格式</param>
        /// <param name="flag">渲染到纹理的渲染标志</param>
        /// <param name="color">渲染到纹理的颜色</param>
        /// <param name="useDepth">渲染到纹理是否使用深度</param>
        /// <param name="screenRatio">渲染到纹理的纵横比</param>
        public void Setup(int width, int height, PixelFormat format, ClearFlag flag, ref Vector4 color, bool useDepth, float screenRatio)
        {
            ICall_RenderToTexture_Setup(this, width, height,(int)format,(uint)flag, ref color,useDepth,screenRatio,0,0,0,0,0);
        }

        /// <summary>
        /// 创建一个用户自定义的渲染到纹理
        /// </summary>
        /// <param name="width">渲染到纹理的宽度</param>
        /// <param name="height">渲染到纹理的高度</param>
        /// <param name="format">渲染到纹理的格式</param>
        /// <param name="flag">渲染到纹理的渲染标志</param>
        /// <param name="color">渲染到纹理的颜色</param>
        /// <param name="useDepth">渲染到纹理是否使用深度</param>
        public void Setup(int width, int height, PixelFormat format, ClearFlag flag, ref Vector4 color, bool useDepth)
        {
            ICall_RenderToTexture_Setup(this, width, height, (int)format, (uint)flag, ref color, useDepth, 0, 0, 0, 0, 0, 0);
        }

        internal override IntPtr GetTextureHandlePtr()
        {
            return ICall_RenderToTexture_GetTextureHandlePtr(this);
        }

        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern private static void ICall_RenderToTexture_Bind(RenderToTexture self);

        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern private static void ICall_RenderToTexture_Release(RenderToTexture self);

        
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern private static void ICall_RenderToTexture_Setup(RenderToTexture self,int width,int height,int pixelFormat
            , uint flag, ref Vector4 color, bool useDepth
			,float screenRatio,int antiAliasQuality,int texLeft,int texRight,int texTop,int texBottom);
         
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        extern private static IntPtr ICall_RenderToTexture_GetTextureHandlePtr(RenderToTexture self);
    }
}
